File: tests/testthat/test-burden.R
test_that("calculate_burden runs end-to-end and returns one row per cell type", {
  data <- list(
    small = sim_cell_type(25, n_genes = 60, seed = 30),
    large = sim_cell_type(150, n_genes = 60, seed = 31)
  )
  res <- calculate_burden(
    data, de_fun = "wilcoxon",
    grid_args = list(n_min = 12, n_points = 3),
    k = 2, k_permute = 2, seed = 100
  )
  expect_s3_class(res, "burden_result")
  expect_equal(nrow(res$burden), 2)
  expect_true(all(c("cell_type", "n_full", "reference_n", "burden", "burden_se") %in%
                    names(res$burden)))
  expect_true(all(res$burden$burden >= 0))
})

test_that("calculate_burden works with adaptive_alpha = FALSE and forced pooling/model_type", {
  data <- list(
    a = sim_cell_type(20, n_genes = 50, seed = 32),
    b = sim_cell_type(80, n_genes = 50, seed = 33)
  )
  res <- calculate_burden(
    data, de_fun = "wilcoxon", model_type = "linear", pooling = "stratified",
    adaptive_alpha = FALSE,
    grid_args = list(n_min = 10, n_points = 3),
    k = 2, k_permute = 0, seed = 101
  )
  expect_equal(res$pooling_used, "stratified")
  expect_true(is.list(res$model) && !inherits(res$model, "burden_model"))
  expect_null(res$alpha_curves)
})

test_that("calculate_burden errors if adaptive_alpha requested without permuted reps", {
  data <- list(a = sim_cell_type(20, n_genes = 30, seed = 34))
  expect_error(
    calculate_burden(data, de_fun = "wilcoxon", adaptive_alpha = TRUE, k_permute = 0),
    "k_permute > 0"
  )
})

test_that("reference_n defaults to the median full cell count across cell types", {
  data <- list(
    a = sim_cell_type(20, n_genes = 30, seed = 35),
    b = sim_cell_type(60, n_genes = 30, seed = 36),
    c = sim_cell_type(100, n_genes = 30, seed = 37)
  )
  res <- calculate_burden(
    data, de_fun = "wilcoxon", adaptive_alpha = FALSE,
    grid_args = list(n_min = 10, n_points = 3),
    k = 2, k_permute = 0, seed = 102
  )
  expect_equal(unique(res$burden$reference_n), 60)
})

File: python/tests/test_burden.py
import numpy as np
import pytest

from burden import calculate_burden

pytestmark = pytest.mark.filterwarnings("ignore")


@pytest.fixture
def data(sim):
    return {
        "small": sim(n_genes=200, n_per_group=25, n_de=15, effect=2.5, seed=1),
        "medium": sim(n_genes=200, n_per_group=100, n_de=15, effect=2.5, seed=2),
        "large": sim(n_genes=200, n_per_group=300, n_de=15, effect=2.5, seed=3),
    }


def _run(data, **kwargs):
    defaults = dict(de_fun="wilcoxon", k=3, k_permute=3, grid_args={"n_points": 5}, seed=11)
    defaults.update(kwargs)
    return calculate_burden(data, **defaults)


def test_default_pooling_is_pooled(data):
    res = _run(data)
    assert res.pooling_used == "pooled"
    # pooled model -> identical burden across cell types at the same reference_n
    assert res.burden["burden"].nunique() == 1


def test_burden_output_columns(data):
    res = _run(data)
    expected = {
        "cell_type",
        "n_full",
        "reference_n",
        "n_degs_observed",
        "burden",
        "burden_se",
        "extrapolated",
    }
    assert expected <= set(res.burden.columns)
    assert len(res.burden) == 3


def test_reference_n_defaults_to_median_full_n(data):
    res = _run(data)
    assert res.burden["reference_n"].iloc[0] == pytest.approx(100)


def test_reference_n_override(data):
    res = _run(data, reference_n=50)
    assert (res.burden["reference_n"] == 50).all()


def test_extrapolation_flag_and_warning(data):
    with pytest.warns(UserWarning, match="extrapolated"):
        res = _run(data, reference_n=1000)
    assert res.burden["extrapolated"].all()


def test_pooling_test_always_computed_even_when_pooled_forced(data):
    res = _run(data, pooling="pooled")
    assert res.pooling_used == "pooled"
    assert res.pooling_test["recommendation"] in ("pooled", "shared_shape", "stratified")


def test_stratified_pooling_gives_per_cell_type_models(data):
    res = _run(data, pooling="stratified")
    assert res.pooling_used == "stratified"
    assert set(res.model.keys()) == set(data.keys())


def test_shared_shape_pooling(data):
    res = _run(data, pooling="shared_shape")
    assert res.pooling_used == "shared_shape"


def test_invalid_pooling_raises(data):
    with pytest.raises(ValueError):
        _run(data, pooling="bogus")


def test_adaptive_alpha_requires_k_permute(data):
    with pytest.raises(ValueError):
        _run(data, adaptive_alpha=True, k_permute=0)


def test_adaptive_alpha_false_skips_alpha_curves(data):
    res = _run(data, adaptive_alpha=False)
    assert res.alpha_curves is None
